### DISCUSSION TITLE
How to get the BLE data to build my own app?

### DISCUSSION THREAD

**Discussion by alexhoopes** [renatobo/bonogps#41] [Q&A] [answered]
I'm working on my own iOS app that tracks 0-60 times and location data. I'm just getting started and was wondering how I can interface with the BLE on this? Does anyone have any direction in what I need to learn to be able to pull the NMEA sentences off via BLE? I'll be able to parse all the data after that, I just can't figure out how to interface with it and receive the sentences.
Thanks!

Edit: I tried looking at the data from a BLE discovery app in the Location and Navigation service but it shows no data. Not sure if I'm looking in the wrong place![image](https://github.com/user-attachments/assets/9a597d41-d50a-4dbb-b59f-a7d222a87e0e)

I also see that nothing is being printed on serial in terms of NMEA sentences, is that supposed to be the case?
![image](https://github.com/user-attachments/assets/4f97b2ba-efff-4719-9544-58cce69c36ac)

I've also tried the following Python to try and read the BLE data but nothing gets printed:
```
import asyncio
from bleak import BleakClient, BleakScanner

# Constants from the BonoGPS code
LOCATION_SERVICE_UUID = "00001819-0000-1000-8000-00805f9b34fb"  # 0x1819 SERVICE_LOCATION_AND_NAVIGATION
LOCATION_CHAR_UUID = "00002A67-0000-1000-8000-00805f9b34fb"    # 0x2A67 LOCATION_AND_SPEED
DEVICE_NAME_PREFIX = "BonoGPS"

async def notification_handler(sender, data):
    """Print raw data from the GPS device"""
    print(f"\nRaw bytes: {data}")
    try:
        # Try to decode as ASCII and print
        ascii_data = data.decode('ascii')
        print(f"ASCII decoded: {ascii_data}")
    except UnicodeDecodeError:
        print("Could not decode as ASCII")
    print("-" * 50)

async def scan_for_device():
    print("Scanning for BonoGPS devices...")
    devices = await BleakScanner.discover()
    
    for device in devices:
        if device.name and DEVICE_NAME_PREFIX in device.name:
            print(f"Found BonoGPS device: {device.name} ({device.address})")
            return device
            
    print("No BonoGPS devices found!")
    return None

async def main():
    try:
        # Scan for device
        device = await scan_for_device()
        if not device:
            return

        print(f"\nConnecting to {device.name}...")
        async with BleakClient(device.address) as client:
            print("Connected!")
            
            # Print services and characteristics for debugging
            print("\nAvailable services:")
            for service in client.services:
                print(f"Service: {service.uuid}")
                for char in service.characteristics:
                    print(f"  Characteristic: {char.uuid}")
                    print(f"    Properties: {char.properties}")
            
            print("\nStarting notifications...")
            await client.start_notify(LOCATION_CHAR_UUID, notification_handler)
            
            print("\nReceiving raw data (Press Ctrl+C to stop)...")
            while True:
                await asyncio.sleep(0.1)

    except KeyboardInterrupt:
        print("\nStopping...")
    except Exception as e:
        print(f"Error: {e}")

if __name__ == "__main__":
    asyncio.run(main())
```
![image](https://github.com/user-attachments/assets/877a98c2-2bd1-4054-aa6f-bba361d0f384)
No data is coming through

@renatobo, any chance I could get some insight here? Thanks for this amazing program by the way :)

**Comment by alexhoopes** [comment]
It seems the GPIO pins were configured wrong and that's why I wasn't getting data. Fixed it! Time to make the app, still, thank you for this program !!

**Reply by alexhoopes** [reply]
Or maybe not, it was working but after a reboot not anymore... serial log is showing BLE buffer saturated, resetting being spammed over and over

**Reply by renatobo** [reply, marked as answer]
I recommend you get things up and running at 1Hz, with one of the applications I list in the documentation. 
Once you know your GPS and esp32 are working correctly, you should build your app.

Also, before building an app that reads GPS data, try broadcasting a simple message with a serial counter and have it read on your app.

Develop and test each step below separately
- GPS -> esp32
- esp32 to BLE stack
- BLE stack to iOS BLE stack
- BLE stack on iOS to app
